soroban-env-host/src/host.rs

```diff
@@ -1659,18 +1666,81 @@ impl VmCallerEnv for Host {
         _vmcaller: &mut VmCaller<Host>,
         k: RawVal,
         v: RawVal,
+        t: StorageType,
+        f: RawVal,
     ) -> Result<Void, HostError> {
-        let key = self.contract_data_key_from_rawval(k)?;
-        let data = LedgerEntryData::ContractData(ContractDataEntry {
-            contract_id: self.get_current_contract_id_internal()?,
-            key: self.from_host_val(k)?,
-            val: self.from_host_val(v)?,
-        });
-        self.0.storage.borrow_mut().put(
-            &key,
-            &Host::ledger_entry_from_data(data),
-            self.as_budget(),
-        )?;
+        let flags = if f.is_void() {
+            0
+        } else {
+            self.u32_from_rawval_input("f", f)?
+        };
+
+        if ((flags as u64) & !MASK_CONTRACT_DATA_FLAGS_V20) != 0 {
+            return Err(self.err(
+                ScErrorType::Value,
+                ScErrorCode::InvalidInput,
+                "invalid flags",
+                &[],
+            ));
+        }
+
+        let storage_type: ContractDataType = t.try_into()?;
+        let key = self.contract_data_key_from_rawval(k, storage_type)?;
+        if self.0.storage.borrow_mut().has(&key, self.as_budget())? {
```

Comment:
minor nit -- read-only borrows can be `.borrow()` not `borrow_mut()`

Reply:
We can't use `borrow` here because `has` and `get` both take a mutable reference to self. I get the following error if I try that. 
```
trait `DerefMut` is required to modify through a dereference, but it is not implemented for `std::cell::Ref<'_, storage::Storage>`
```

Reply:
ok